File: static/pick_score.js
// Single source of truth for the draft pick-score FORMULA.
//
// Mirrors utils/pick_score.py::compute_pick_score exactly. The draft room
// (browser) and the Teams-page /api/draft-grades (Python) both grade picks with
// this same math, so a pick's grade can't differ between the two surfaces.
// tests/test_pick_score_parity.py runs this (via Node) against the Python copy
// over random inputs and fails CI on any divergence.
//
// This is the pure math only - pick QUALITY, nothing time-sensitive. Callers
// gather the inputs (value, vor, tier, adp, need, ppg, ...). Live-draft timing
// (survival to the next pick, redraft handcuff insurance) lives entirely in the
// draft-room decision layer (draft_board_core.js decisionScore), so the timing
// signal is counted exactly once and never runs through the depth/relabel
// shaping below.
(function (root, factory) {
  var api = factory();
  if (typeof module === 'object' && module.exports) module.exports = api;
  root.BRPickScore = api;
})(typeof self !== 'undefined' ? self : this, function () {
  function clamp01(x) { return x < 0 ? 0 : (x > 1 ? 1 : x); }

  // Rookie momentum down-weighted 0.06 -> 0.03 (freed weight to value/adp) after
  // a 509-team backtest. Rookie & startup youth set to the validated 0.10 winner
  // (600-league multi-year). Rookie 0.16 -> 0.10 (freed mass to tier/ppg);
  // startup 0.05 -> 0.10 (taken from ADP). Need weights are unchanged. Keep in
  // lockstep with PS_WEIGHTS in utils/pick_score.py. Redraft model value is
  // deliberately DB-only (not an ADP blend); its explicit market weight starts
  // at .18 and decays with round depth, with the released weight redistributed
  // exactly. Live timing (survival, handcuff, late-round upside) is NOT here.
  var WEIGHTS = {
    rookie:  { vor: 0.06, value: 0.20, adp: 0.30, tier: 0.18, need: 0.05, youth: 0.10, mom: 0.03, ppg: 0.13 },
    redraft: { vor: 0.15, value: 0.25, adp: 0.18, tier: 0.12, need: 0.09, youth: 0.00, mom: 0.03, ppg: 0.22 },
    startup: { vor: 0.07, value: 0.24, adp: 0.25, tier: 0.15, need: 0.09, youth: 0.10, mom: 0.03, ppg: 0.12 },
  };
  var AGE_PEAKS = { RB: 24, WR: 27, TE: 27, QB: 29 };
  var CORE = { QB: 1, RB: 1, WR: 1, TE: 1 };

  // o: { pos, value, vor, tier, age, rankChange7d, avgPick, pickNo, maxVal,
  //      draftType, isSf, needRaw, qbCount, totalPicks, numTeams, ppgNorm,
  //      ppr, tep, passTd, isTierCliff }
  // Returns an integer 0-100.
  function computePickScore(o) {
    var pos = (o.pos || '').toUpperCase();
    var value = o.value != null ? +o.value : 0;
    var vor = o.vor != null ? +o.vor : null;
    var age = o.age != null ? +o.age : null;
    var rankChange7d = o.rankChange7d != null ? +o.rankChange7d : null;
    var avgPick = o.avgPick != null ? +o.avgPick : null;
    var maxVal = o.maxVal != null ? +o.maxVal : 0;
    var needRaw = o.needRaw != null ? +o.needRaw : 0;
    var totalPicks = o.totalPicks != null ? +o.totalPicks : 0;
    var pickNo = o.pickNo != null ? +o.pickNo : 0;
    var tier = o.tier;
    var ppr = o.ppr != null ? +o.ppr : 1.0;
    var tep = o.tep != null ? +o.tep : 0.0;
    var passTd = o.passTd != null ? +o.passTd : 4.0;

    var dbValueNorm = (maxVal && maxVal > 0) ? clamp01(value / maxVal) : 0;
    // Keep model/database value independent from the explicit market component.
    // A near-zero value also caps selected-only ADP below.
    var adpUntrusted = dbValueNorm < 0.05;
    var valueNorm = dbValueNorm;
    // Scarcity residual: share of this player's own value above replacement.
    // Same-position players with similar value get nearly the same residual.
    var vorNorm;
    if (vor != null && value > 0) vorNorm = clamp01(Math.max(vor, 0) / value);
    else if (vor != null) vorNorm = 0;
    else vorNorm = valueNorm * 0.8;

    var adpVal;
    if (avgPick != null) {
      var rel = (pickNo - avgPick) / Math.max(avgPick, 1.5);
      if (rel >= 0.5) adpVal = 1.0;
      else if (rel >= -0.3) adpVal = 0.5 + rel;
      else adpVal = Math.max(0, 0.2 + rel * 0.25);
      if (avgPick <= 8) adpVal = Math.max(adpVal, clamp01(0.5 + (8 - avgPick) / 16));
      if (adpUntrusted) adpVal = Math.min(adpVal, 0.5);
    } else {
      adpVal = 0.5;
    }

    var tierScore = tier ? clamp01((10 - Math.min(tier, 9)) / 9) : valueNorm;
    if (o.isTierCliff) tierScore = clamp01(tierScore + 0.15);

    // Need ramps in a touch earlier than before (/10 vs /12) so roster
    // construction starts to matter before the mid rounds, not only after.
    var needRamp = clamp01((pickNo - 1) / 10.0);
    var need = (1 - needRamp) * 0.5 + needRamp * needRaw;

    var youth = 0.5;
    if (age != null && CORE[pos]) {
      var peak = AGE_PEAKS[pos] || 27;
      youth = clamp01((peak - age + 4) / 8);
    }
    var mom = clamp01((rankChange7d || 0) / 20 + 0.5);
    var ppgN = o.ppgNorm != null ? +o.ppgNorm : valueNorm;

    var w = WEIGHTS[o.draftType] || WEIGHTS.startup;
    if (o.draftType === 'redraft') {
      var teams0 = Math.max(1, +o.numTeams || 12);
      var round0 = pickNo ? Math.floor((pickNo - 1) / teams0) + 1 : 1;
      var adpFactor = Math.max(0.15, 1 - Math.max(0, round0 - 2) * 0.075);
      var freed = w.adp * (1 - adpFactor), base = w.vor + w.value + w.tier + w.need + w.ppg;
      w = Object.assign({}, w);
      w.adp *= adpFactor;
      ['vor','value','tier','need','ppg'].forEach(function(k){ w[k] += freed * w[k] / base; });
    }
    var s = w.vor * vorNorm + w.value * valueNorm + w.adp * adpVal
          + w.tier * tierScore + w.need * need + w.youth * youth
          + w.mom * mom + (w.ppg || 0) * ppgN;

    if (!o.isSf && pos === 'QB') {
      var teams = o.numTeams ? +o.numTeams : 12;
      var round = pickNo ? Math.floor((pickNo - 1) / Math.max(teams, 1)) + 1 : 1;
      var qc = o.qbCount || 0;
      var pen = 1.0;
      if (qc >= 1) {
        // Overfill: a 2nd+ QB in 1QB only carries real opportunity cost early;
        // by the late rounds a backup QB is a normal pick, so it tapers out.
        pen = round <= 3 ? 0.30 : (round <= 6 ? 0.60 : (round <= 9 ? 0.85 : 1.0));
        if (qc >= 2) pen *= 0.7;
      } else if (o.draftType === 'redraft') {
        // First QB, redraft only: QB is streamable in 1QB, so a startable QB
        // shouldn't leapfrog clear starters / ADP-fallers in the early-mid
        // rounds. Tapers to none by the late rounds (grabbing a QB is fine
        // then). Dynasty/startup keep full QB value - young QBs are long-term
        // assets, not a streamable weekly slot.
        pen = round <= 3 ? 0.82 : (round <= 6 ? 0.90 : (round <= 9 ? 0.97 : 1.0));
      }
      s *= pen;
    }

    // Redundancy: a pick at a skill position already stocked to its realistic
    // depth target (needRaw == 0) is a bench body at a full spot while other
    // starting needs may remain. Penalize it, hardest at a true single-starter
    // slot (1-TE, or starterSlots <= 1) and in the early rounds. Need math is
    // unchanged — only the single-vs-multi classification uses roster slots.
    if (needRaw <= 0 && (pos === 'RB' || pos === 'WR' || pos === 'TE')) {
      var teams2 = o.numTeams ? +o.numTeams : 12;
      var rd2 = pickNo ? Math.floor((pickNo - 1) / Math.max(teams2, 1)) + 1 : 1;
      var single = (o.starterSlots != null && isFinite(+o.starterSlots))
        ? (+o.starterSlots <= 1)
        : (pos === 'TE');
      var rp = rd2 <= 3 ? (single ? 0.55 : 0.82)
             : rd2 <= 6 ? (single ? 0.72 : 0.90)
             : rd2 <= 9 ? (single ? 0.86 : 0.96)
             : 1.0;
      s *= rp;
    }

    if (tep && tep > 0 && pos === 'TE') s *= (1 + 0.12 * tep);
    // Six-point passing TDs raise the weekly ceiling and replacement advantage
    // of quarterbacks. Keep this modest in 1QB and let the existing SF/VOR logic
    // handle format scarcity rather than double-counting it here.
    if (pos === 'QB' && passTd >= 6) s *= 1.06;
    if (pos === 'WR' || pos === 'TE') { if (ppr != null && ppr >= 1) s *= 1.02; }
    else if (pos === 'RB' && ppr != null && ppr <= 0) s *= 1.03;

    if (totalPicks && totalPicks > 1 && pickNo) {
      var depth = Math.min(0.98, (pickNo - 1) / totalPicks);
      var par = Math.max(0.40, 1.0 - depth * 0.44);
      s = s / par;
    }

    // Display relabel (monotonic): everything above is the backtested ranking
    // and is left untouched. This only stretches the near-ceiling band so the
    // best pick's 0-100 number differentiates instead of clustering at ~97.
    // Scores under the knee (~85) are unchanged; above it the curve is steeper
    // than 1:1, so a truly elite pick pulls toward 100 while a merely-good "best
    // available" reads lower. Monotonic => it never changes which pick ranks
    // higher, only the label. Keep identical to utils/pick_score.py.
    var d = clamp01(s);
    var _knee = 0.85;
    if (d > _knee) {
      var _t = (d - _knee) / (1 - _knee);
      d = _knee + _t * _t * (1 - _knee);
    }
    return Math.floor(d * 100 + 0.5);
  }

  // Effective starters per position from a league's roster slot counts, used to
  // anchor VOR and PPG replacement levels. Shared so the draft room and the
  // server derive the SAME replacement index from the SAME roster (the WR/RB
  // gap between them was the main cause of mismatched grades). Splits SF half to
  // QB and FLEX half each to RB/WR, matching the draft room's computeReplacement.
  //   counts: { QB, SF, RB, WR, TE, FLEX }  ->  { QB, RB, WR, TE } (floats)
  function starterCounts(counts) {
    var c = counts || {};
    var n = function (k) { return +c[k] || 0; };
    return {
      QB: n('QB') + n('SF') * 0.5,
      RB: n('RB') + n('FLEX') * 0.5 + n('RB_WR') * 0.5 + n('RB_TE') * 0.5,
      WR: n('WR') + n('FLEX') * 0.5 + n('RB_WR') * 0.5 + n('WR_TE') * 0.5,
      TE: n('TE') + n('WR_TE') * 0.5 + n('RB_TE') * 0.5,
    };
  }

  return { computePickScore: computePickScore, starterCounts: starterCounts };
});

File: utils/pick_score.py
"""Pure draft pick-score computation.

Extracted from app.py so the pick-scoring engine can be unit-tested without the
pandas/DB stack. ``compute_pick_score`` blends DB dynasty value, ADP, tier,
positional need, youth, momentum and production into a 0-100 score, mirroring
the client-side Draft Room. Pure — reuses ``clamp01`` from utils.draft_grade.
"""
from __future__ import annotations

import math

from utils.draft_grade import clamp01

# Component weights per draft type (approximately normalized within each row).
PS_WEIGHTS = {
    "rookie":  {"vor": 0.06, "value": 0.18, "adp": 0.29, "tier": 0.12, "need": 0.05, "youth": 0.24, "mom": 0.06, "ppg": 0.05},
    "redraft": {"vor": 0.10, "value": 0.24, "adp": 0.33, "tier": 0.08, "need": 0.07, "youth": 0.00, "mom": 0.03, "ppg": 0.18},
    "startup": {"vor": 0.07, "value": 0.24, "adp": 0.30, "tier": 0.12, "need": 0.09, "youth": 0.10, "mom": 0.03, "ppg": 0.10},
}
PS_AGE_PEAKS = {"RB": 24, "WR": 27, "TE": 27, "QB": 29}


def ps_tier_of(value: float, thresholds: list):
    """Tier number for a dynasty value given gap-significance thresholds (1 = elite)."""
    if not thresholds:
        return None
    for i, thr in enumerate(thresholds):
        if value >= thr:
            return i + 1
    return len(thresholds) + 1


def starter_counts(counts: dict) -> dict:
    """Mirror of static/pick_score.js `starterCounts`; pinned by the parity test.
    Effective starters per position from roster slot counts (SF split half to QB,
    FLEX split half each to RB/WR), so the server's VOR/PPG replacement levels
    match the draft room's computeReplacement instead of a hardcoded guess."""
    c = counts or {}

    def n(k):
        try:
            return float(c.get(k) or 0)
        except (TypeError, ValueError):
            return 0.0

    return {
        "QB": n("QB") + n("SF") * 0.5,
        "RB": n("RB") + n("FLEX") * 0.5,
        "WR": n("WR") + n("FLEX") * 0.5,
        "TE": n("TE"),
    }


def compute_pick_score(*, pos, value, vor, tier, age, rank_change_7d,
                       avg_pick, pick_no, max_val, draft_type, is_sf,
                       need_raw, qb_count, total_picks=None, num_teams=None,
                       ppg_norm=None, ppr=1.0, tep=0.0, is_tier_cliff=False,
                       survival_adj=0.0, handcuff=False) -> int:
    """Mirror of static/pick_score.js `computePickScore`; the two are pinned
    identical by tests/test_pick_score_parity.py. ``survival_adj`` and
    ``handcuff`` are the live-draft-only timing terms (default off); the grade
    never passes them, which is why the Teams-page grade and the Draft Room
    grade agree. Do not edit this without editing the JS (and vice versa)."""
    pos = (pos or "").upper()
    # DB-sourced numbers arrive as decimal.Decimal; coerce to float so they mix
    # with the float weights below (Decimal * float raises TypeError).
    value = float(value) if value is not None else 0.0
    vor = float(vor) if vor is not None else None
    age = float(age) if age is not None else None
    rank_change_7d = float(rank_change_7d) if rank_change_7d is not None else None
    avg_pick = float(avg_pick) if avg_pick is not None else None
    max_val = float(max_val) if max_val is not None else 0.0
    need_raw = float(need_raw) if need_raw is not None else 0.0
    total_picks = float(total_picks) if total_picks is not None else 0.0
    db_value_norm = clamp01(value / max_val) if max_val and max_val > 0 else 0.0
    # Blend DB dynasty value with ADP-implied quality so market consensus
    # prevents DB value gaps (especially new rookies) from dragging the score
    # unfairly low when ADP says the player is a legitimate round-2/3 pick.
    if avg_pick is not None and total_picks > 0:
        adp_qual_norm = clamp01(1.0 - avg_pick / total_picks)
        value_norm = db_value_norm * 0.35 + adp_qual_norm * 0.65
    else:
        value_norm = db_value_norm
    vor_norm = clamp01(vor / max(max_val, 1)) if vor is not None else value_norm * 0.8

    # ADP component: proportional gap so a 2-pick fall from ADP 2 == a 10-pick
    # fall from ADP 20, with an elite-ADP floor for top-8 players.
    if avg_pick is not None:
        gap = pick_no - avg_pick
        rel = gap / max(avg_pick, 1.5)
        if rel >= 0.5:
            adp_val = 1.0
        elif rel >= -0.3:
            adp_val = 0.5 + rel
        else:
            adp_val = max(0.0, 0.2 + rel * 0.25)
        if avg_pick <= 8:
            adp_val = max(adp_val, clamp01(0.5 + (8 - avg_pick) / 16))
    else:
        adp_val = 0.5

    tier_score = clamp01((10 - min(tier, 9)) / 9) if tier else value_norm
    # Tier-cliff boost: position scarcity when this player's tier is drying up
    # (<=2 left in the bucket). Mirrors the Draft Room's isTierCliff() bump.
    if is_tier_cliff:
        tier_score = clamp01(tier_score + 0.15)

    need_ramp = clamp01((pick_no - 1) / 12.0)
    need = (1 - need_ramp) * 0.5 + need_ramp * need_raw

    youth = 0.5
    if age is not None and pos in ("RB", "WR", "TE", "QB"):
        peak = PS_AGE_PEAKS.get(pos, 27)
        youth = clamp01((peak - age + 4) / 8)

    mom = clamp01((rank_change_7d or 0) / 20 + 0.5)

    # Production: position-normalized PPG. Missing data falls back to value_norm
    # so a player isn't penalized for absent projections (mirrors the Draft Room).
    ppg_n = ppg_norm if ppg_norm is not None else value_norm

    w = PS_WEIGHTS.get(draft_type, PS_WEIGHTS["startup"])
    s = (w["vor"] * vor_norm + w["value"] * value_norm + w["adp"] * adp_val
         + w["tier"] * tier_score + w["need"] * need + w["youth"] * youth
         + w["mom"] * mom + w.get("ppg", 0.0) * ppg_n)

    # Live-draft survival/opportunity-cost term (0 for grading).
    if survival_adj:
        s += float(survival_adj)

    # QB overfill (1QB only): a second QB only carries real opportunity cost in
    # the early rounds. By the late rounds a backup QB is a normal pick, so the
    # penalty tapers out (mirrors the Draft Room).
    if not is_sf and pos == "QB" and qb_count >= 1:
        _teams = int(num_teams) if num_teams else 12
        _round = (int(pick_no) - 1) // max(_teams, 1) + 1 if pick_no else 1
        if _round <= 3:
            _pen = 0.30
        elif _round <= 6:
            _pen = 0.60
        elif _round <= 9:
            _pen = 0.85
        else:
            _pen = 1.0
        if qb_count >= 2:
            _pen *= 0.7
        s *= _pen

    # Live-draft redraft handcuff term (False for grading).
    if handcuff:
        s = min(1.0, s + 0.15)

    # Scoring-format adjustments: shift toward the build the league's scoring
    # rewards. Mirrors the Draft Room's scoringCfg() multipliers exactly.
    if tep and tep > 0 and pos == "TE":
        s *= (1 + 0.12 * tep)
    if pos in ("WR", "TE"):
        if ppr is not None and ppr >= 1:
            s *= 1.02
    elif pos == "RB" and ppr is not None and ppr <= 0:
        s *= 1.03

    # Depth normalization: re-anchor the 0-100 scale to what's achievable at this
    # pick slot so late-round picks aren't unfairly buried (mirrors the Draft Room).
    if total_picks and total_picks > 1 and pick_no:
        _depth = min(0.98, (float(pick_no) - 1) / float(total_picks))
        _par = max(0.40, 1.0 - _depth * 0.44)
        s = s / _par

    return int(math.floor(clamp01(s) * 100 + 0.5))  # round-half-up, matching JS
